Mathlib/Data/Finset/Density.lean

```diff
@@ -0,0 +1,217 @@
+  rw [← dens_union_of_disjoint sdiff_disjoint, sdiff_union_self_eq_union]
+
+lemma dens_sdiff_comm [IsCancelAdd 𝕜] (h : card s = card t) : dens[𝕜] (s \ t) = dens (t \ s) :=
+  add_left_injective (dens t) $ by
+    simp_rw [dens_sdiff_add_dens, union_comm s, ← dens_sdiff_add_dens, dens, h]
+
+@[simp]
+lemma dens_sdiff_add_dens_inter (s t : Finset α) : dens[𝕜] (s \ t) + dens (s ∩ t) = dens s := by
+  rw [← dens_union_of_disjoint (disjoint_sdiff_inter _ _), sdiff_union_inter]
+
+@[simp]
+lemma dens_inter_add_dens_sdiff (s t : Finset α) : dens[𝕜] (s ∩ t) + dens (s \ t) = dens s := by
+  rw [add_comm, dens_sdiff_add_dens_inter]
+
+lemma dens_filter_add_dens_filter_not_eq_dens
+    (p : α → Prop) [DecidablePred p] [∀ x, Decidable (¬p x)] :
+    dens[𝕜] (s.filter p) + dens (s.filter fun a ↦ ¬ p a) = dens s := by
+  classical
+  rw [← dens_union_of_disjoint (disjoint_filter_filter_neg ..), filter_union_filter_neg_eq]
+
+end Lattice
+
+section CharZero
+variable [CharZero 𝕜]
+
+@[simp] lemma dens_eq_zero : dens[𝕜] s = 0 ↔ s = ∅ := by
+  simp (config := { contextual := true }) [dens, Fintype.card_eq_zero_iff, eq_empty_of_isEmpty]
+
+lemma dens_ne_zero : dens[𝕜] s ≠ 0 ↔ s.Nonempty := dens_eq_zero.not.trans nonempty_iff_ne_empty.symm
+
+protected alias ⟨_, Nonempty.dens_ne_zero⟩ := dens_ne_zero
+
+variable [Nonempty α]
+
+@[simp] lemma dens_univ : dens[𝕜] (univ : Finset α) = 1 := by simp [dens, card_univ]
+
+@[simp] lemma dens_eq_one : dens[𝕜] s = 1 ↔ s = univ := by
+  simp [dens, div_eq_one_iff_eq, card_eq_iff_eq_univ]
+
+lemma dens_ne_one : dens[𝕜] s ≠ 1 ↔ s ≠ univ := dens_eq_one.not
+
+end CharZero
+
+end Semifield
+
+section LinearOrderedSemifield
+variable [LinearOrderedSemifield 𝕜] {s t : Finset α} {a b : α}
+
+@[simp] lemma dens_nonneg : 0 ≤ dens[𝕜] s := by unfold dens; positivity
+
+lemma dens_le_dens (h : s ⊆ t) : dens[𝕜] s ≤ dens t :=
+  div_le_div_of_nonneg_right (mod_cast card_mono h) $ by positivity
+
+lemma dens_lt_dens (h : s ⊂ t) : dens[𝕜] s < dens t :=
+  div_lt_div_of_pos_right (mod_cast card_strictMono h) $ by
+    cases isEmpty_or_nonempty α
+    · simp [Subsingleton.elim s t, ssubset_irrfl] at h
+    · positivity
+
+@[mono] lemma dens_mono : Monotone (dens : Finset α → 𝕜) := fun _ _ ↦ dens_le_dens
+@[mono] lemma dens_strictMono : StrictMono (dens : Finset α → 𝕜) := fun _ _ ↦ dens_lt_dens
+
+lemma dens_map_le [Fintype β] (f : α ↪ β) : dens[𝕜] (s.map f) ≤ dens s := by
+  cases isEmpty_or_nonempty α
+  · simp [Subsingleton.elim s ∅]
+  simp_rw [dens, card_map]
+  gcongr
+  · positivity
+  · positivity
+  · exact Fintype.card_le_of_injective _ f.2
+
+section Lattice
+variable [DecidableEq α]
+
+lemma dens_union_le (s t : Finset α) : dens[𝕜] (s ∪ t) ≤ dens s + dens t :=
+  dens_union_add_dens_inter (𝕜 := 𝕜) s t ▸ le_add_of_nonneg_right dens_nonneg
+
+lemma dens_le_dens_sdiff_add_dens : dens[𝕜] s ≤ dens (s \ t) + dens t := by
+  simp_rw [dens, ← add_div, ← Nat.cast_add]
+  gcongr
+  · positivity
+  · exact card_le_card_sdiff_add_card
```

Comment:
```suggestion
  dens_sdiff_add_dens (s := s) (t := t) (𝕜 := 𝕜) ▸ dens_le_dens (subset_union_left s t)
```

Maybe some of the arguments of `dens_sdiff_add_dens` should be explicit to make this even cleaner.